### PR TITLE
fix panic in UpdateNetworkContainers

### LINKED ISSUE #2754: Panic in github.com/docker/swarm/cluster.(*Engine).UpdateNetworkContainers
I am running docker legacy swarm manager as a service in a docker swarm mode cluster. The legacy swarm manager is restarted every minute after the following panic:

```
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | panic: assignment to entry in nil map
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | 
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | goroutine 274 [running]:
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | panic(0xc2b7c0, 0xc4203448f0)
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | 	/usr/local/go/src/runtime/panic.go:500 +0x1a1
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | github.com/docker/swarm/cluster.(*Engine).UpdateNetworkContainers(0xc4203849a0, 0x0, 0x0, 0x0, 0x0, 0x0)
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | 	/go/src/github.com/docker/swarm/cluster/engine.go:1025 +0x3ca
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | github.com/docker/swarm/cluster.(*Engine).refreshLoop(0xc4203849a0)
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | 	/go/src/github.com/docker/swarm/cluster/engine.go:949 +0x605
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | created by github.com/docker/swarm/cluster.(*Engine).ValidationComplete
swarm-master.1.uvuzsekbrpxt@ukwtsvulx157.redacted.com    | 	/go/src/github.com/docker/swarm/cluster/engine.go:365 +0xab
```

Despite this error using the docker legacy swarm inside the docker swarm mode is working fine for my use case (starting jenkins one-off build slaves).

swarm version:
```
Server:
 Version:      swarm/1.2.7
 API version:  1.22 (minimum version )
 Go version:   go1.7.1
 Git commit:   2005aec
 Built:        2017-06-30T22:13:10+0000
 OS/Arch:      linux/amd64
 Experimental: false
```

docker version:
```
Server:
 Version:      17.05.0-ce
 API version:  1.29 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   89658be
 Built:        Thu May  4 22:06:25 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

### PR DESCRIPTION
**Pull request by allencloud**
this PR fixes #2754 

I think when a network is created without connecting to any containers. The containers field is a nil. Thus when we need to assign one element to this field, it is better for us to initialize this map first.

What I did:

1. choose to fail fast to simplify code;
2. make a map ptr when assignment needed. 

ping @dongluochen @nishanttotla 

Signed-off-by: allencloud <allen.sun@daocloud.io>

### CODE REVIEW

File: cluster/engine.go [hunk 1]
```diff
@@ -996,38 +996,42 @@ func (e *Engine) UpdateNetworkContainers(containerID string, full bool) error {
...
+			// extract ip addresses
+			ipv4address := ""
+			ipv6address := ""
+			if n.IPAddress != "" {
+				ipv4address = n.IPAddress + "/" + strconv.Itoa(n.IPPrefixLen)
+			}
+			if n.GlobalIPv6Address != "" {
+				ipv6address = n.GlobalIPv6Address + "/" + strconv.Itoa(n.GlobalIPv6PrefixLen)
+			}
+			// udpate network information
```
Comment (nishanttotla) [inline cluster/engine.go]: nit: `udpate` -> `update`

**Comment by nishanttotla** [comment]
Thanks for the quick fix @allencloud. 

I think we'll have to do another release to ensure this fix is picked up 😞 

I'll start it tomorrow, once this PR is merged.

File: cluster/engine.go (hunk 1 above)
Comment (allencloud) [inline cluster/engine.go]: Thanks a lot. Have updated this.

**Comment by allencloud** [comment]
Yeah, I think another release is reasonable. @nishanttotla

**Comment by tisoft** [comment]
Any news about the release? I would like to test the fix on our cluster :)

**Comment by nishanttotla** [comment]
@tisoft I will start the release and most likely create an RC in a day or so that you can test and report back :)

**Comment by nishanttotla** [comment]
@tisoft please check 1.2.8-rc1: https://github.com/docker/swarm/releases/tag/v1.2.8-rc1

Let us know if this fixes your problem. The official image should be available soon once the library is updated.

**Comment by tisoft** [comment]
Looking good! Seems the issue is fixed. Thanks!

**Comment by nishanttotla** [comment]
@tisoft great! We'll do the main release soon then.

**Pull request merged**